internal/internal_worker.go

```diff
@@ -765,58 +774,68 @@ func (ae *activityExecutor) GetFunction() interface{} {
 	return ae.fn
 }
 
+func (ae *activityExecutor) initializedActivityInterceptor(ctx context.Context) (ActivityInboundCallsInterceptor, error) {
+	// Initial interceptor is the one that reflectively calls the function
+	var interceptor ActivityInboundCallsInterceptor = &activityReflectInterceptor{fn: ae.fn}
+	// Chain interceptors in reverse order
+	interceptors := getInterceptorsFromActivityCtx(ctx)
+	for i := len(interceptors) - 1; i >= 0; i-- {
+		interceptor = interceptors[i].InterceptActivity(ctx, interceptor)
+	}
+	// Initialize interceptor
+	return interceptor, interceptor.Init(ctx)
+}
+
 func (ae *activityExecutor) Execute(ctx context.Context, input *commonpb.Payloads) (*commonpb.Payloads, error) {
-	fnType := reflect.TypeOf(ae.fn)
-	var args []reflect.Value
 	dataConverter := getDataConverterFromActivityCtx(ctx)
 
-	// activities optionally might not take context.
-	if fnType.NumIn() > 0 && isActivityContext(fnType.In(0)) {
-		args = append(args, reflect.ValueOf(ctx))
+	// Prepare interceptor, panicking on error
+	// TODO(cretz): Panicking acceptable? This is what workflow interceptors do.
+	interceptor, err := ae.initializedActivityInterceptor(ctx)
+	if err != nil {
+		panic(err)
 	}
 
-	decoded, err := decodeArgs(dataConverter, fnType, input)
+	decoded, err := decodeArgsToValues(dataConverter, reflect.TypeOf(ae.fn), input)
 	if err != nil {
 		return nil, fmt.Errorf(
 			"unable to decode the activity function input payload with error: %w for function name: %v",
 			err, ae.name)
 	}
-	args = append(args, decoded...)
 
-	fnValue := reflect.ValueOf(ae.fn)
-	retValues := fnValue.Call(args)
-	return validateFunctionAndGetResults(ae.fn, retValues, dataConverter)
+	// TODO(cretz): Make sure getActivityEnvironmentFromCtx(ctx).activityType.Name == ae.name
+	// TODO(cretz): Acceptable to only use string name like workflow interceptors instead of ActivityType struct?
+	ret := interceptor.ExecuteActivity(ctx, ae.name, decoded...)
```

Comment:
You may want to double check this with some other tests. I think there may very well be some extant bugs around this stuff.

In my (possibly wrong) understanding, we allow users to call activities three ways:
* By string name
* By reference to a method on an actual instantiated struct
* By reference to a method on a nil pointer whose type is an activity struct (something like `(*ActivityStruct).method()`

No. 3 should be equivalent to No. 2 since there are not any non-static values that the method could operate on. (Also I think option 3 is crazytown and we likely shouldn't even allow it, but, different convo).

No 2 on the other hand does make sense and is kinda just a different way to enclose state that changes between scheduled activities.

So, probably want to double check those possibilities (and entirely possible you end up fixing a bug in the process)

Reply:
> You may want to double check this

To clarify, when you say "this" are you referring to the "activity type"? I will add tests assert activity type in those three ways. FWIW, the integration tests are invoking this for all activities (which includes those 3 from my reading).

Reply:
Yeah, that's what I meant by this. Thanks for double checking 👍